engine architecture two sections (#2)
YetAnotherSpieskowcy/Carcassonne-Documentation: engine_architecture.tex

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam consectetur dignissim urna, a rhoncus dolor venenatis adipiscing. Phasellus viverra aliquam fringilla. Praesent mattis diam id ipsum placerat a laoreet diam tempus. Quisque eu leo ut sapien ultricies gravida ac sed turpis. Donec nibh enim, pretium a sagittis et, varius non nulla. Vivamus enim augue, condimentum eget convallis hendrerit, mattis ut lectus. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Proin semper libero augue. Praesent auctor ligula sit amet quam molestie adipiscing. Mauris vehicula congue mollis. Duis tincidunt interdum purus quis congue. Integer vel orci mauris. Curabitur rutrum tincidunt mi ut elementum. Nunc non massa nunc. Donec ultricies ipsum ac risus sollicitudin porta. Phasellus elementum odio et leo consectetur consectetur. 

\subsection{Components of a solution}

> % Opisanie elementów z jakich się składa rozgrywka m.in.: kafelki i ich elementy, gracze, a także uwzględnione rozszerzenia gry Carcassonne.

Reviewer: Discussion topic: is the description in the comment what we actually intended to include in this section? It's not consistent with the contents of the progress report last semester but it seems consistent with the description that we included in the article draft.
The text that's already here is fine, it's just a matter of whether there isn't something else that we wanted to include that we might otherwise miss.

Author: well, I followed the article draft topics, however after checking the semester report I believe my section describes more or less the idea of our solution. I do not think that huge class diagram would fit into the article. Moreover there is no sense describing all structures, but the most crucial ones. The only things I omited were stacks, scoreReport and  logger. Logger could be mentioned in the first paragraph.

The game Carcassone is played by placing tiles \cite{CarcassoneRules}. Each tile is represented by a set of features on it. All features are defined by their type, such as road, city, field or monastery, a modifier that modifies the calculation process, and the sides which indicate on which edges the feature is present. In this project there is only one modifier used - the shield. The engine was designed to accommodate game extensions. 

A feature's side is a byte sized value, in which each bit defines whether the feature is present on a specific half of the tile's edge as shown on Figure \ref{fig:SIDES}. Features such as roads, are represented by a whole edge formed by combining two sides e.g., the TopLeft and TopRight sides.
\begin{figure}
    \centering
    \begin{tikzpicture}
        %top line
        \draw [blue] (-3,3) -- (0,3);
        \draw [red] (0,3) -- (3,3);
        %right line
        \draw [blue] (3,3) -- (3,0);
        \draw [red] (3,0) -- (3,-3);

>         %bottom line
>         \draw [red] (3,-3) -- (0,-3);
>         \draw [blue] (0,-3) -- (-3,-3);

Reviewer 2: ```suggestion
        %bottom line
        \draw [blue] (3,-3) -- (0,-3);
        \draw [red] (0,-3) -- (-3,-3);
```
(other lines also have colors in this order, and now the image is point-symmetrical)

        %left line
        \draw [blue] (-3,-3) -- (-3,0);
        \draw [red] (-3,0) -- (-3,3);
        
        \node at (-1.5, 2.7) {TopLeft};
        \node at (1.5, 2.7) {TopRight};
        
        \node [rotate=90] at (2.8, 1.5) {RightTop};
        \node [rotate=90] at (2.8, -1.5) {RightBottom};
        
        \node at (-1.5, -2.8) {BottomLeft};
        \node at (1.5, -2.8) {BottomRight};
        
	% czy może obrót 90stopni?
        \node [rotate=270] at (-2.8, 1.5) {LeftTop};
        \node [rotate=270] at (-2.8, -1.5) {LeftBottom};
    \end{tikzpicture}
    \caption{Side representation} \label{fig:SIDES}
\end{figure}

> The tiles described do not contain information about their position or meeple placement. This information is located in the PlacedTile struct. The board is represented as a slice of placed tiles ordered by their placement, moreover it contains the used tile set, for verification whether the tile a player wants to place is legal. Additionally, board includes a city manager used for merging cities, scoring and checking for legal meeples placement in the cities.

Reviewer: > The board is represented as a slice of placed tiles ordered by their placement

If we want to be truthful, this is actually not true - the order of tiles is dependent on the order of tiles in the tile set *template*.

Author: corrected

> A single game is represented by a board, shuffled deck of tiles, and players taking their turns. Each player knows their score, and the number of meeples of each type available to them.

Reviewer 2: > [...] and players taking their turns. [...]

I think "and players partaking in it" or "and a ?slice? of players" would be more fitting. "Taking their turns" is an action so it is not included in the game representation per se.

Author: changed for "and players partaking in it"

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam consectetur dignissim urna, a rhoncus dolor venenatis adipiscing. Phasellus viverra aliquam fringilla. Praesent mattis diam id ipsum placerat a laoreet diam tempus. Quisque eu leo ut sapien ultricies gravida ac sed turpis. Donec nibh enim, pretium a sagittis et, varius non nulla. Vivamus enim augue, condimentum eget convallis hendrerit, mattis ut lectus. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Proin semper libero augue. Praesent auctor ligula sit amet quam molestie adipiscing. Mauris vehicula congue mollis. Duis tincidunt interdum purus quis congue. Integer vel orci mauris. Curabitur rutrum tincidunt mi ut elementum. Nunc non massa nunc. Donec ultricies ipsum ac risus sollicitudin porta. Phasellus elementum odio et leo consectetur consectetur. 

\subsection{Communication within a system}
% Opisanie komunikacji pomiędzy silnikiem gry, a sieciami neuronowymi uczonymi na karcie graficznej. Aktualizacja stanu gry w zależności od akcji podejmowanych przez agenta.

> The game engine is capable of handling multiple game instances, by using golang pipelines and goroutines \cite{GolangPipeline}, which allows for many agents to play and learn simultaneously. The engine provides several requests necessary for the agents, to retrieve information about the game. There are five such requests:

Reviewer 2: ```suggestion
The game engine is capable of handling multiple game instances, by using Go pipelines and goroutines \cite{GolangPipeline}, which allows for many agents to play and learn simultaneously. The engine provides several requests necessary for the agents, to retrieve information about the game. There are five such requests:
```

Probably an unnecessary nitpick, but "golang" is an informal name, and the official name is just Go

> The game engine is capable of handling multiple game instances, by using golang pipelines and goroutines \cite{GolangPipeline}, which allows for many agents to play and learn simultaneously. The engine provides several requests necessary for the agents, to retrieve information about the game. There are five such requests:

Reviewer 2: ```suggestion
The game engine is capable of handling multiple game instances, by using golang pipelines and goroutines \cite{GolangPipeline}, which allows for many agents to play and learn simultaneously. The engine provides several request types necessary for the agents, to retrieve information about the game. There are five such requests:
```

Author: changed

\begin{itemize}
    \item CloneGame - used for creating a separate clone of a game. It allows the agent to expand the game tree, and each clone can be modified by the agent, by simulating its next potential move.

>     \item GetRemainingTiles -  returns all available tiles that have not yet been placed. With this data, the agent can expand the game tree for each possible move, that might occur. The returned tiles come with information about their probability.

Reviewer 2: ```suggestion
    \item GetRemainingTiles -  returns all available tiles that have not yet been placed. With this data, the agent can expand the game tree for each possible move that might occur. The returned tiles come with information about their probability.
```

>     \item GetLegalMoves - returns all legal moves for a given tile. It considers where tile can be placed and where the possible meeple placement on it.

Reviewer 2: ```suggestion
    \item GetLegalMoves - returns all legal moves for a given tile. It considers where the tile can be placed and where are the possible meeple placements on it.
```

    \item GetMidGameScore - returns the information about the score for all players, calculated as if the game has just finished.
    \item PlayTurn - allows the agent to play a turn using a given tile, and modifies the game state. It returns the new state of the game 
\end{itemize}

% coś o tym że  agent może budować drzewo jak w algorytmie alphago, ale troche nie wiem jak ten akapit rozwinąć.
With these requests, the agent is capable of building its own game tree, as in the AlphaGo algorithm \cite{AlphaGoAlgorithm}.

> To provide the neural network with the game state, a serialization process is required. Serialized game contains all players' statistics (meeple counts, score)  and all tiles on the board. The placed tiles on the board are represented in the binary format to ensure a fixed memory size of 8 bytes. The slice of binary tiles consists of all tiles in placement order. The tiles in this array that are not yet placed are set to zero.

Reviewer 2: ```suggestion
To provide the neural network with the game state, a serialization process is required. Serialized game contains all players' statistics (meeple counts, score) and all tiles on the board. The placed tiles on the board are represented in the binary format to ensure a fixed memory size of 8 bytes. The slice of binary tiles consists of all tiles in placement order. The tiles in this array that are not yet placed are set to zero.
```

> To provide the neural network with the game state, a serialization process is required. Serialized game contains all players' statistics (meeple counts, score) and all tiles on the board. The placed tiles on the board are represented in the binary format to ensure a fixed memory size of 8 bytes. The slice of binary tiles consists of all tiles in placement order. The tiles in this array that are not yet placed are set to zero.

Reviewer: > The slice of binary tiles consists of all tiles in placement order.

If we want to be truthful, this is actually not true - the order of tiles is dependent on the order of tiles in the tile set *template*.

Author: corrected
